### PR TITLE
[MRG] Linear Neural Networks/Implementation of Softmax Regression from Scratch

### PR DESCRIPTION
**Pull request by xiaotinghe**

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-zh/PR-557/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-557/

**Comment by mli** [comment]
Job d2l-zh/PR-557/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-557/

**Comment by mli** [comment]
Job d2l-zh/PR-557/3 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-557/

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
-# Softmax 回归的从零开始实现
+# softmax回归的从零开始实现
 :label:`sec_softmax_scratch`
 
-就像我们从头开始实现线性回归一样，我们认为 softmax 回归也是类似的基础，你应该知道如何自己实现它的血腥细节。我们将使用刚刚在 :numref:`sec_fashion_mnist` 中引入的时尚 MNist 数据集，设置批量大小为 256 的数据迭代器。
+就像我们从零开始实现线性回归一样，我们认为softmax回归也是重要的基础，你应该知道如何自己实现它的细节。我们使用刚刚在 :numref:`sec_fashion_mnist` 中引入的Fashion-MNIST数据集，并设置数据迭代器的批量大小为256。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:4]: “我们认为softmax回归也是重要的基础，你应该知道如何自己实现它的细节” 缺乏逻辑信号词，不连贯

“我们认为softmax回归也是重要的基础，因此你应该知道如何自己实现它的细节”

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 2]
```diff
@@ -32,9 +32,9 @@ train_iter, test_iter = d2l.load_data_fashion_mnist(batch_size)
 
 ## 初始化模型参数
 
-与我们的线性回归样本一样，这里的每个样本都将用固定长度向量表示。原始数据集中的每个样本都是 $28 \times 28$ 图像。在本节中，我们将拼合每个图像，将它们视为长度为 784 的矢量。未来，我们将讨论利用图像空间结构的更复杂的策略，但现在我们将每个像素位置视为另一个特征。
+这里的每个样本都用固定长度向量表示。原始数据集中的每个样本都是 $28 \times 28$ 的图像。在本节中，我们将展平每个图像，将它们视为长度为784的向量。在以后的章节中，我们将讨论利用图像空间结构的更复杂策略，但现在我们将每个像素位置视为一个特征。
 
-回想一下，在 softmax 回归中，我们的输出数量与类一样多。因为我们的数据集有 10 个类，所以我们的网络输出维度为 10。因此，我们的权重将构成一个 $784 \times 10$ 矩阵，偏差将构成一个 $1 \times 10$ 行向量。与线性回归一样，我们将使用高斯噪声初始化我们的权重 `W`，我们的偏差将初始值 0。
+回想一下，在softmax回归中，我们的输出数与类别一样多。因为我们的数据集有10个类别，所以网络输出维度为 10。因此，权重将构成一个 $784 \times 10$ 的矩阵，偏差将构成一个 $1 \times 10$ 的行向量。与线性回归一样，我们将使用正态分布初始化我们的权重 `W`，偏差初始化为0。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:37]: 输出与类别一样多

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 3]
```diff
@@ -65,9 +65,9 @@ W = tf.Variable(tf.random.normal(shape=(num_inputs, num_outputs),
 b = tf.Variable(tf.zeros(num_outputs))
 ```
 
-## 定义软最大操作
+## 定义softmax操作
 
-在实现 softmax 回归模型之前，让我们简要地回顾一下总和运算符如何沿着张量中的特定维度工作，如 :numref:`subseq_lin-alg-reduction` 和 :numref:`subseq_lin-alg-non-reduction` 所述。给定矩阵 `X`，我们可以对所有元素进行总和（默认情况下）或仅对同一轴中的元素进行总和，即同一列（轴 0）或同一行（轴 1）。请注意，如果 `X` 是一个具有形状（2,3）的张量，并且我们对列进行总和，则结果将是一个具有形状（3，）的向量。在调用和运算符时，我们可以指定保留原始张量中的轴数，而不是折叠我们所汇总的尺寸。这将导致形状（1，3）的二维张量。
+在实现softmax回归模型之前，让我们简要地回顾一下sum运算符如何沿着张量中的特定维度工作，如 :numref:`subseq_lin-alg-reduction` 和 :numref:`subseq_lin-alg-non-reduction` 所述。给定一个矩阵`X`，我们可以对所有元素求和（默认情况下），也可以只求同一个轴上的元素，即同一列（轴0）或同一行（轴1）。如果 `X` 是一个形状为（2,3）的张量，我们对列进行总和，则结果将是一个具有形状（3，）的向量。当调用sum操作符时，我们可以指定保持在原始张量的轴数，而不折叠求和的维度。这将产生一个具有形状（1，3）的二维张量。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:70]: sum -> `sum`

Comment (astonzhang) [reply chapter_linear-networks/softmax-regression-scratch.md:70]: 对列进行总和
->对列进行求和

Comment (astonzhang) [reply chapter_linear-networks/softmax-regression-scratch.md:70]: 操作符 -》 运算符

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 4]
```diff
@@ -81,31 +81,35 @@ X = d2l.tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
 d2l.reduce_sum(X, 0, keepdims=True), d2l.reduce_sum(X, 1, keepdims=True)
 ```
 
-我们现在已经准备好实施 softmax 操作了。回想一下，softmax 由三个步骤组成：i) 我们对每个项进行指数（使用 `exp`）; ii) 我们对每行进行总和（我们在批量中每个样本中有一行）以获得每个样本的归一化量; iii) 我们将每行除以其标准化常量，确保结果总和为 1。在查看代码之前，让我们回顾一下这是如何表达的方程式：
+我们现在已经准备好实现softmax操作了。回想一下，softmax 由三个步骤组成：
+i） 我们对每个项求幂（使用“exp”）；
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:85]: `exp`

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 5]
```diff
@@ -81,31 +81,35 @@ X = d2l.tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
...
+ii）我们对每一行求和（小批量中每个样本是一行），得到每个样本的归一化常数；
+iii）我们将每一行除以其归一化常数，确保结果和为1。
+在查看代码之前，让我们回顾一下这个表达式：
 
 $$
 \mathrm{softmax}(\mathbf{X})_{ij} = \frac{\exp(\mathbf{X}_{ij})}{\sum_k \exp(\mathbf{X}_{ik})}.
 $$
 
-分母或标归一化量，有时也称为 * 分区函数 *（其对数称为日志分区函数）。该名称的起源来自 [统计物理学](https://en.wikipedia.org/wiki/Partition_function_(statistical_mechanics)，其中一个相关的方程对粒子集合上的分布进行建模。
+分母或归一化常数，有时也称为*配分函数*（其对数称为对数-配分函数）。该名称的起源来自 [statistical physics](https://en.wikipedia.org/wiki/Partition_function_(statistical_mechanics))中一个模拟粒子群的分布的方程。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:94]: 统计物理学

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 6]
```diff
@@ -121,23 +125,23 @@ X_prob = softmax(X)
...
 ```
 
-## 定义损耗函数
+## 定义损失函数
 
-接下来，我们需要实现 :numref:`sec_softmax` 中引入的交叉熵损失函数。这可能是所有深度学习中最常见的损失函数，因为目前分类问题远远超过回归问题。
+接下来，我们需要实现 :numref:`sec_softmax` 中引入的交叉熵损失函数。这可能是深度学习中最常见的损失函数，因为目前分类问题的数量远远超过回归问题。
 
-回想一下，交叉熵采用分配给真实标签的预测概率的负对数似然。我们可以通过单个运算符选择所有元素，而不是使用 Python for 循环迭代预测（这往往效率低下）。下面，我们创建了一个玩具数据 `y_hat`，其中包含 3 个类的预测概率的 2 个示例。然后我们选择第一个例子中第一个类的概率和第二个样本中第三个类的概率。
+回顾一下，交叉熵采用真实标签的预测概率的负对数似然。我们不需要使用Python for 循环迭代预测（这往往是低效的）。我们可以通过一个操作符选择所有元素。下面，我们创建一个演示数据 `y_hat`，其中包含2个样本在3个类别的预测概率。然后我们选择第一个样本中第一个类的概率和第二个样本中第三个类的概率。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:144]: “Python的for循环”
这样避免连续两个英文单词

Comment (astonzhang) [reply chapter_linear-networks/softmax-regression-scratch.md:144]: 操作符 -》 运算符

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 7]
```diff
@@ -174,30 +178,30 @@ cross_entropy(y_hat, y)
 
 ## 分类准确性
 
-给定预测概率分布 `y_hat`，只要我们必须输出硬预测，我们通常选择具有最高预测概率的类。事实上，许多应用需要我们做出选择。Gmail 必须将电子邮件分类为 “主”、“社交”、“更新” 或 “论坛”。它可能会在内部估计概率，但在一天结束时，它必须在类中选择一个。
+在给定预测概率分布 `y_hat`时，当我们必须输出硬预测时，我们通常选择预测概率最高的类。许多应用都要求我们做出选择。如Gmail必须将电子邮件分为“Primary（主要）”、“Social（社交）”、“Updates（更新）”或“Forums（论坛）”。它可能在内部估计概率，但最终它必须在类中选择一个。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:181]: 硬预测（hard prediction）

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 8]
```diff
@@ -174,30 +178,30 @@ cross_entropy(y_hat, y)
...
 
-当预测与标签分类 `y` 一致时，它们是正确的。分类准确率是所有正确预测的一部分。虽然直接优化准确率性可能很困难（这是不可区分的），但它通常是我们最关心的性能衡量标准，我们在训练分类器时几乎总是会报告它。
+当预测与标签分类 `y` 一致时，它们是正确的。分类准确率即正确预测数量与总预测数量之比。虽然直接优化准确率可能很困难（因为准确率的计算不可导），但准确率通常是我们最关心的性能衡量标准，我们在训练分类器时几乎总是会报告它。
 
-为了计算准确率，我们执行以下操作。首先，如果 `y_hat` 是矩阵，我们假设第二个维度存储每个类的预测分数。我们使用 `argmax` 通过每行中最大条目的索引获取预测类。然后我们将预测类与地面真实 `y` 元素进行比较。由于等式运算符 `==` 对数据类型很敏感，因此我们将 `y_hat` 的数据类型转换为与 `y` 的数据类型相匹配。结果是一个包含 0（false）和 1（true）条目的张量。获取总和会产生正确预测的数量。
+为了计算准确率，我们执行以下操作：首先，如果 `y_hat` 是矩阵，第二个维度存储每个类的预测分数。我们使用 `argmax` 获得每行中最大元素的索引来获得预测类别。然后我们将预测类别与真实 `y` 元素进行比较。由于等式运算符 `==` 对数据类型很敏感，因此我们将 `y_hat` 的数据类型转换为与 `y` 的数据类型一致。结果是一个包含 0（错）和 1（对）的张量。进行求和会得到正确预测的数量。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:185]: 冒号改句号。

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 9]
```diff
@@ -240,16 +245,16 @@ class Accumulator:  #@save
...
 ```{.python .input}
 #@tab all
 evaluate_accuracy(net, test_iter)
 ```
 
-## 培训
+## 训练
 
-如果您阅读我们在 :numref:`sec_linear_scratch` 中的线性回归的实现，softmax 回归的训练循环应该看起来非常熟悉。在这里，我们重构实现以使其可重复使用。首先，我们定义一个函数来训练一个迭代周期（周期)。请注意，`updater` 是更新模型参数的常规函数，它接受批量次大小作为参数。它可以是 `d2l.sgd` 函数的包装器，也可以是框架的内置优化函数。
+如果你看过 :numref:`sec_linear_scratch` 中的线性回归的实现，softmax回归的训练循环应该看起来非常熟悉。在这里，我们重构训练循环的实现以使其可重复使用。首先，我们定义一个函数来训练一个迭代周期。请注意，`updater` 是更新模型参数的常用函数，它接受批量大小作为参数。它可以是包装的`d2l.sgd`函数，也可以是框架的内置优化函数。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:257]: 包装 -》 封装

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 10]
```diff
@@ -373,7 +378,7 @@ class Animator:  #@save
         display.clear_output(wait=True)
 ```
 
-然后，以下训练函数在通过 `train_iter` 访问的训练数据集上训练一个模型 `net`，该数据集由 `num_epochs` 指定。在每个迭代周期（周期) 结束时，通过 `test_iter` 访问的测试数据集对模型进行评估。我们将利用 `Animator` 类来可视化培训进度。
+接下来的训练函数会在`train_iter` 访问的训练数据集上训练一个模型`net`，将会运行多个迭代周期(由`num_epochs`指定)。在每个迭代周期结束时，利用 `test_iter` 访问的测试数据集对模型进行评估。我们将利用 `Animator` 类来可视化训练进度。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:381]: “ 接下来的训练函数会在`train_iter` 访问的训练数据集上训练一个模型`net`，将会运行多个迭代周期(由`num_epochs`指定)。”

不通顺

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 11]
```diff
@@ -441,18 +446,18 @@ def predict_ch3(net, test_iter, n=6):  #@save
 predict_ch3(net, test_iter)
 ```
 
-## 摘要
+## 总结
 
-* 借助 softmax 回归，我们可以训练多类分类的模型。
-* softmax 回归的训练循环与线性回归中的训练循环非常相似：检索和读取数据，定义模型和损失函数，然后使用优化算法训练模型。正如你很快就会发现的，大多数常见的深度学习模型都有类似的培训程序。
+* 借助 softmax 回归，我们可以训练多分类的模型。
+* softmax 回归的训练循环与线性回归中的训练循环非常相似：读取数据、定义模型和损失函数，然后使用优化算法训练模型。正如你很快就会发现的，大多数常见的深度学习模型都有类似的训练过程。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:452]: 正如你很快就会发现的那样，

File: chapter_linear-networks/softmax-regression-scratch.md [hunk 12]
```diff
@@ -441,18 +446,18 @@ def predict_ch3(net, test_iter, n=6):  #@save
...
 
 ## 练习
 
-1. 在本节中，我们直接实现了基于 softmax 运算的数学定义的 softmax 函数。这可能会导致什么问题？提示：尝试计算 $\exp(50)$ 的大小。
-1. 本节中的函数 `cross_entropy` 是根据交叉熵损失函数的定义实现的。这个实现可能有什么问题？提示：考虑对数的域。
+1. 在本节中，我们直接实现了基于数学定义softmax运算的softmax函数。这可能会导致什么问题？提示：尝试计算 $\exp(50)$ 的大小。
```
Comment (astonzhang) [inline chapter_linear-networks/softmax-regression-scratch.md:456]: `softmax`函数

**Comment by mli** [comment]
Job d2l-zh/PR-557/4 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-557/

**Comment by mli** [comment]
Job d2l-zh/PR-557/5 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-557/

**Comment by mli** [comment]
Job d2l-zh/PR-557/6 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-557/

**Comment by mli** [comment]
Job d2l-zh/PR-557/7 is complete. 
Check the results at http://preview.d2l.ai/d2l-zh/PR-557/
